public/app/plugins/datasource/grafana/components/QueryEditor.tsx

```diff
@@ -245,7 +251,8 @@ export class QueryEditor extends PureComponent<Props, State> {
   }
 
   render() {
-    const query = defaults(this.props.query, defaultQuery);
+    const { query } = this.props;
```

Comment:
can we not apply the defaults immutably here in render instead of doing it mutably in the constructor? or does the rest of the component rely on the mutation of props.query?

Reply:
@torkelo we can, the only reason i didn't want to do that was that it then introduces a disconnect between the json and what's displayed on screen.

## when switching to the --Grafana-- datasource and not modifying anything

applying it mutably in the constructor:
```json
      ...
      "targets": [
        {
          "queryType": "randomWalk",
          "refId": "A"
        }
      ]
      ...
```

applying it immutably in render:
```json
      ...
      <no targets key at all>
      ...
```

in both cases everything works as expected in the UI. tbh i'm not sure which is worse, the disconnect or the mutating. 😅  happy to defer to your wisdom 🧙

Reply:
defaults should not be saved anyway (ideally), so that is not a big issue

Reply:
cool, that makes things easier. i've updated to apply them immutably in the render method 👍